provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -286,72 +318,181 @@ func uploadImage(
 	}
 	defer f.Close()
 
-	// Transfer image upload progress to the UpdateProgress function.
-	progressChan := make(chan progress.Report)
-	progressSink := progressUpdater{
-		ch:     progressChan,
-		update: updateProgress,
-		action: fmt.Sprintf("uploading %s", item.Path),
+	// Transfer upload progress to the updateStatus function.
+	statusUpdater := statusUpdater{
+		ch:       make(chan progress.Report),
+		clock:    clock,
+		logger:   logger,
+		update:   updateStatus,
+		action:   fmt.Sprintf("uploading %s", item.Path),
+		interval: updateStatusInterval,
 	}
-	go progressSink.loop()
-	defer progressSink.done()
+	go func() {
+		defer statusUpdater.tomb.Done()
+		statusUpdater.tomb.Kill(statusUpdater.loop())
+	}()
 
+	// Update the lease periodically.
+	leaseUpdater := leaseUpdater{
+		ch:                  make(chan progress.Report),
+		clock:               clock,
+		logger:              logger,
+		ctx:                 ctx,
+		leaseUpdaterContext: leaseUpdaterContext,
+	}
+	go func() {
+		defer leaseUpdater.tomb.Done()
+		leaseUpdater.tomb.Kill(leaseUpdater.loop())
+	}()
+
+	// Upload.
 	opts := soap.Upload{
-		Method:        "POST",
-		Type:          "application/x-vnd.vmware-streamVmdk",
 		ContentLength: item.Size,
-		Progress:      &progressSink,
+		Progress:      progress.Tee(&statusUpdater, &leaseUpdater),
+	}
+	if item.Create {
+		opts.Method = "PUT"
+		opts.Headers = map[string]string{"Overwrite": "t"}
+	} else {
+		opts.Method = "POST"
+		opts.Type = "application/x-vnd.vmware-streamVmdk"
 	}
-	if err := client.Upload(f, targetURL, &opts); err != nil {
-		return errors.Annotatef(err, "uploading %s to %s", filepath.Base(item.Path), targetURL)
+	doUpload := func() error {
+		// NOTE(axw) client.Upload is not cancellable,
+		// as there is no way to inject a context. We
+		// should send a patch to govmomi to make it
+		// cancellable.
+		return errors.Trace(client.Upload(f, targetURL, &opts))
+	}
+
+	var site catacomb.Catacomb
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &site,
+		Work: doUpload,
+		Init: []worker.Worker{&statusUpdater, &leaseUpdater},
+	}); err != nil {
+		return errors.Trace(err)
 	}
-	return nil
+	return site.Wait()
 }
 
-type progressUpdater struct {
-	tomb   tomb.Tomb
-	ch     chan progress.Report
-	update func(string)
-	action string
+type statusUpdater struct {
```

Comment:
should this be made a simple worker? which would remove a little of the boilerplate